### PR TITLE
Fix Broken Admin Stats Graphic on Desktop App Boot

### PR DESCRIPTION
**Pull request by nsakaimbo**
### Description

Attempting to fix the broken admin bar stats graphic of the "Stats" menu item on the main Calypso landing page.

### Context

- At application boot, the graphic is displayed as a broken link due to a forbidden URL request:

<p align="center">
admin-bar-broken
</p>

Console error: `status 403 (forbidden): https://<site_name>.wordpress.com/wp-includes/charts/admin-bar-hours-scale-2x.php?masterbar=1&s=<site_id>`

On a refresh triggered through the console, the graphic displays correctly.

<p align="center">
admin-bar-not-broken
</p>

Closer inspection of headers sent with the bad/good requests indicate that the bad request is missing the `wordpress_logged_in` cookie in the `Cookie` header. TL;DR: it appears the `wordpress_logged_in` cookie has not yet been parsed and set in Electron by the time the request for the `admin-bar-hours` graphic is made. (However, the cookie _is_ present when the renderer is refreshed with `CMD + R` from the Developer Console.

At present, cookies from Calypso/renderer process are relayed via a `user-auth` IPC event ([link](https://github.com/Automattic/wp-calypso/blob/3a779a0ca38129848941bc4bbe6f4d38486214e3/client/desktop/lib/cookie-auth/index.js#L101-L126)). However, triggering this IPC event isn't conditional on Calypso having completely finished its boot sequence (where things like user bootstrapping take place). Ideally the sequence-of-events here should be deterministic and Calypso should be fully booted after the environment is set and Electron's main process has completed the user authentication step (i.e. writing cookies to the user's window's `webSession`). 

### Fix

Ensure that Calypso rendering takes place after Electron main process cookie-handling is complete.

### To Test

Download and run the build artifacts associated for this PR. Boot the application and note that the graphic for the `Stats` menu item is displayed on first boot.

<p align="center">
Screen Shot 2020-10-27 at 11 19 43 AM
</p>

### CODE REVIEW

File: client/server/pages/index.js [hunk 1]
```diff
@@ -109,6 +112,24 @@ function setupLoggedInContext( req, res, next ) {
 	next();
 }
 
+async function setupDesktopUserAuth( req, res, next ) {
```
Comment (nsakaimbo) [inline client/server/pages/index.js:115]: I'm not very familiar with the boot logic and not sure if I'm on the right path here.

Problem: this doesn't actually work (app boots to a blank white screen). Also the conditional import of `electron/ipc` is laughable here, although that being said I'm not entirely sure how to insert the `user-auth` IPC event within the existing boot sequence.

Also, one interesting fact about the Electron side of things is that the main window is created _after_ the Calypso server has completed initializing ([link](https://github.com/Automattic/wp-calypso/blob/6e1f5f5ed209c216a8751ec28031199b7e6accf8/client/desktop/server/index.js#L151)). Cookies sent over IPC are parsed and saved to the web session of an already-created window object ([link](https://github.com/Automattic/wp-calypso/blob/6e1f5f5ed209c216a8751ec28031199b7e6accf8/client/desktop/lib/cookie-auth/index.js#L73-L96)). 

TL;DR If the Electron window is created _after_ the server finishes initializing, and if the window is required to parse and set cookies relayed by IPC, not sure if/how this general approach will work. 🤔

**Review by wp-desktop — DISMISSED** [review]
WordPress Desktop CI Failure for job "wp-desktop-mac".

@nsakaimbo please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/929153). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

File: client/lib/desktop/index.js [hunk 2]
```diff
@@ -204,7 +205,7 @@ const Desktop = {
 		debug( 'Sending logged-in = ' + status );
 
 		ipc.send( 'user-login-status', status );
-		ipc.send( 'user-auth', user(), oAuthToken.getToken() );
+		return ipc.invoke( 'user-auth', user(), oAuthToken.getToken() );
```
Comment (nsakaimbo) [inline client/lib/desktop/index.js:208]: Context: the `client/lib/desktop/index.js` `init()` method is invoked by Calypso's boot process in `client/boot/common.js` ([link](https://github.com/Automattic/wp-calypso/blob/91d5605b0ea4211a5752c26373888aec90233cef/client/boot/common.js#L384)):

```
if ( config.isEnabled( 'desktop' ) ) {
		require( 'lib/desktop' ).default.init();
}
```

Electron's `ipc.invoke` API returns a promise that is resolved when the main process responds to the invoke call ([Electron docs](https://www.electronjs.org/docs/api/ipc-main#ipcmainhandlechannel-listener)).

**Problem**: despite this being apparently sound in theory, this still doesn't solve the problem. When the application is booted, the `admin-bar-hours-scale` graphic is fails to load with a `403 forbidden` status, indicating some sort of authentication issue. A Developer Console reload fixes the graphic. In fact, logging to the developer console indicates that the request for the `admin-bar-hours-scale` asset is _still_ made prior to this init method returning:

<p align="center">
asset-request-init-bug
</p>

The problem appears to be some strange timing-of-events issue that occurs only when a new application window is being created.

File: client/desktop/lib/cookie-auth/index.js [hunk 3]
```diff
@@ -70,50 +67,54 @@ function parseCookie( cookieStr ) {
...
+	ipc.handleOnce( 'user-auth', async function ( event, user, token ) {
+		log.info( `Handling 'user-auth' IPC event, setting session cookies...` );
 
-	ipc.on( 'user-auth', function ( event, user, token ) {
 		if ( user && user.data ) {
 			const userData = user.data;
-			if ( currentRequest && currentRequest.username === userData.username ) {
-				// already authing
-				return;
+			const response = await authorize( userData.username, token );
```
Comment (nsakaimbo) [inline client/desktop/lib/cookie-auth/index.js:104]: Refactor: use promise-based flow instead of callbacks (easier to reason about).

File: client/lib/desktop/index.js (hunk 2 above)
Comment (nsakaimbo) [reply client/lib/desktop/index.js:208]: Update: based on a suggestion by @belcherj, I refactored to block the render on the user authentication step in the main process. This appears to be working, and e2e tests are passing. See https://github.com/Automattic/wp-calypso/pull/46621/commits/0771f880f4f6e8e652a1981d266529da86f04d9b.

Still a little nervous as hopefully nothing else in the app is broken as a result - will be testing this patch a little more.

File: client/boot/common.js [hunk 4]
```diff
@@ -427,13 +427,32 @@ const boot = ( currentUser, registerRoutes ) => {
...
 		// Isomorphic sections will take care of rendering their `<Layout>` themselves.
-		if ( ! document.getElementById( 'primary' ) ) {
+		if ( ! isDesktop && ! document.getElementById( 'primary' ) ) {
 			renderLayout( reduxStore );
+
+			page.start( { decodeURLComponents: false } );
 		}
 
-		page.start( { decodeURLComponents: false } );
+		if ( isDesktop && loggedIn ) {
```
Comment (nsakaimbo) [inline client/boot/common.js:441]: Taking another look at this, I'm actually a little surprised this refactor works. 

Unless I'm interpreting this incorrectly. this will make it so if (1) we are in a Desktop context, and (2) the user is logged in, then (3) rendering will be blocked on the `cookie-auth-complete` event.

However, the `cookie-auth-complete` event is only fired once (on the window's creation). So I'm confused why this would work after the fact, e.g. when a user navigates to a different page. Wouldn't this check be triggered by each page navigation event, or am I not understanding how this client logic works?

cc: @belcherj

Comment (nsakaimbo) [reply client/boot/common.js:441]: My guess is this `render` logic is fired only on application boot. From that point on, the client takes over rendering. Which is the only way the above would make sense.

If this logic was fired for each page navigation, then this probably wouldn't work as expected.

**Review by wp-desktop — CHANGES_REQUESTED** [review]
WordPress Desktop CI Failure for job "wp-desktop-mac".

@nsakaimbo please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/935103). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

File: client/boot/common.js [hunk 5]
```diff
@@ -427,13 +427,36 @@ const boot = ( currentUser, registerRoutes ) => {
...
+				// If we're not in a WP-Desktop context, render.
+				render();
+			} else if ( loggedIn ) {
+				// WP-Desktop: logged in
+				// Wait on cookie-authentication
+				const ipc = require( 'electron' ).ipcRenderer;
+
+				ipc.on( 'cookie-auth-complete', function () {
+					debug( 'Desktop cookies set, rendering main layout...' );
+					render();
```
Comment (belcherj) [inline client/boot/common.js:452]: Add a timeout to ensure we get some sort of render if this does not fire.

**Review by belcherj — CHANGES_REQUESTED** [review]
Please add some way for render to occur if cookie-auth-complete does not fire. Better to have broken admin stats than no render.

File: client/boot/common.js (hunk 5 above)
Comment (nsakaimbo) [reply client/boot/common.js]: Added timeout handling in e3a2ed791611fc1d5fd2a8eccdea9bb9423138b2.

File: client/desktop/lib/cookie-auth/index.js [hunk 6]
```diff
@@ -70,49 +67,56 @@ function parseCookie( cookieStr ) {
...
-				count--;
-				if ( count === 0 ) {
-					if ( onComplete ) onComplete();
-				}
-			} );
+			window.webContents.session.cookies.set( cookie );
+			count--;
+			if ( count === 0 ) {
+				return resolve( true );
+			}
```
Comment (jsnajdr) [inline client/desktop/lib/cookie-auth/index.js:93]: This is strange: the `cookies.set` function is asynchronous and accepts a completion callback. Now this change removes the callback and doesn't wait for `cookies.set` to finish.

Previously, `onComplete` was called only after all the `cookie.set` calls finished. Now everything is called synchronously and it's not even clear why there is a `new Promise()`. There's no async code left.

File: client/boot/common.js [hunk 7]
```diff
@@ -427,13 +438,42 @@ const boot = ( currentUser, registerRoutes ) => {
...
+
+		const renderDesktopPromise = () => {
+			return new Promise( function ( resolve ) {
+				const ipc = require( 'electron' ).ipcRenderer;
+
+				ipc.on( 'cookie-auth-complete', function () {
+					debug( 'Desktop cookies set, rendering main layout...' );
+					render();
+					resolve();
+				} );
```
Comment (jsnajdr) [inline client/boot/common.js:457]: The best place for waiting for `cookie-auth-complete` is the `bootApp` function at the end of this file. The code would be:
```js
await user.initialize();
if ( config.isEnabled( 'desktop' ) ) {
  await waitForCookieAuth();
}
boot( user, registerRoutes );
```
This way, we wait for the user object to be available and for the cookies to be set before continuing with the boot sequence.

Comment (nsakaimbo) [reply client/boot/common.js:457]: That would be a much more compact way to do it, except this approach would be a non-trivial refactor given the `boot` function is what triggers the cookie authentication logic in the Main process. In sequence:

1. `bootApp` calls `boot`
2. `boot` calls `setupMiddlewares` ([link](https://github.com/Automattic/wp-calypso/blob/d682d34136b055b7c392ab3586f94c77de6a6411/client/boot/common.js#L424))
3. `setupMiddlewares` calls `require( 'lib/desktop' ).default.init()` ([link](https://github.com/Automattic/wp-calypso/blob/d682d34136b055b7c392ab3586f94c77de6a6411/client/boot/common.js#L383-L385))
4. Within `require( 'lib/desktop' ).default.init()`, there is a call to `sendUserLoginStatus`, which sends the IPC `user-auth` event ([link](https://github.com/Automattic/wp-calypso/blob/d682d34136b055b7c392ab3586f94c77de6a6411/client/lib/desktop/index.js#L83)).
5. The main process responds to `user-auth` IPC event by fetching and setting cookies to the Window's webContents session ([link](https://github.com/Automattic/wp-calypso/blob/d682d34136b055b7c392ab3586f94c77de6a6411/client/desktop/lib/cookie-auth/index.js#L101)).

The TL;DR here is that the trigger for cookie authentication happens within the `boot` function, making this approach really tricky.

I actually attempted this refactor in 1d21ee449e4e4f5af87cb508baa5c084ff78d15e, and the refactor was non-trivial. I ended up reverting as it wasn't quite working and failed with a timeout.

Happy to revisit this for a refactor later, but in the mean time we're shooting to get this patch into our upcoming release scheduled for Monday.

File: client/desktop/lib/cookie-auth/index.js (hunk 6 above)
Comment (nsakaimbo) [reply client/desktop/lib/cookie-auth/index.js:93]: When debugging these boot/auth flows in the Desktop app, I found the nested callbacks really hard to navigate and reason about. I took the liberty of flattening the callbacks used in this file to make things a little easier to grok going forward.

Especially for Main process logic, this is also aligned with a number of Electron APIs which have migrated to using promises instead of callbacks in recent release. Funnily enough, the `cookies.set` call is promise-based as well ([docs](https://www.electronjs.org/docs/api/cookies#cookiessetdetails)) (although seems like there's a callback variant to support legacy usage). Interestingly prefixing with `await` ends up breaking things (awaits indefinitely for some reason).

For the time being will fall back to the callback variant of this method, although makes me a little nervous that things may silently break when the callback API deprecation is final.

**Review by belcherj — APPROVED** [review]
Walked through the PR. We can address additional refactors in another PR.

File: client/boot/common.js (hunk 7 above)
Comment (jsnajdr) [reply client/boot/common.js:464]: > Happy to enhance/tweak this refactor later, but in the mean time we're shooting to get this patch into our upcoming release scheduled for Monday.

That's fair -- I'm two weeks late with my review, so I can't expect you to wait for me being 100% happy with the PR before shipping it 🙂 Let's improve both the Pinghub and this PR's code in follow-ups. I believe there's plenty of accidental complexity that can be removed.

File: client/boot/common.js [hunk 8]
```diff
@@ -437,10 +433,55 @@ const boot = ( currentUser, registerRoutes ) => {
...
+
+		return Promise.race( [ promise, timeout ] );
+	};
+
+	const renderPromise = () => {
+		return new Promise( function ( resolve ) {
+			const sendUserAuth = () => {
+				debug( 'Sending user info to desktop...' );
+				ipc.send( 'user-auth', user, getToken() );
+			};
```
Comment (jsnajdr) [inline client/boot/common.js:457]: You can call `ipc.send( 'user-auth' )` here directly instead of creating a function and call it later.

It's safe to add the `ipc.on( 'cookie-auth-complete' )` listener after the `ipc.send( 'user-auth' )` already finished. The only thing that needs to be guaranteed is that both calls are done synchronously in the same event loop tick.

`ipc.send( 'user-auth' )` doesn't send the IPC message immediately, before returning. It merely posts a task to the event loop. And that task is performed after the current task (execute a JS function) is complete.

Almost all async APIs in browser and Node.js work like this.

File: client/boot/common.js [hunk 9]
```diff
@@ -437,10 +433,55 @@ const boot = ( currentUser, registerRoutes ) => {
...
+				clearTimeout( id );
+				reject( `Request timed out in ${ ms } ms` );
+			}, ms );
+		} );
+
+		return Promise.race( [ promise, timeout ] );
+	};
+
+	const renderPromise = () => {
+		return new Promise( function ( resolve ) {
```
Comment (jsnajdr) [inline client/boot/common.js:453]: This doesn't need to be wrapped in a function. Just assigning to a variable works, too:
```js
const renderPromise = new Promise( ( resolve ) => {
  // ...
} );
return promiseTimeout( timeoutMs, renderPromise );
```

File: client/boot/common.js [hunk 10]
```diff
@@ -437,10 +433,55 @@ const boot = ( currentUser, registerRoutes ) => {
...
+
+export const bootApp = async ( appName, registerRoutes ) => {
 	const user = userFactory();
-	user.initialize().then( () => {
-		debug( `Starting ${ appName }. Let's do this.` );
-		boot( user, registerRoutes );
-	} );
+	await user.initialize();
+	if ( config.isEnabled( 'desktop' ) ) {
+		await waitForCookieAuth( user );
```
Comment (jsnajdr) [inline client/boot/common.js:483]: If waiting for the `cookie-auth-complete` event times out, then `waifForCookieAuth` returns a rejected promise and the `await` throws an exception. That avoids the entire boot sequence, nothing is ever rendered, and adding the timeout didn't really improve the user's situation.

It would be better to time out gracefully: return a resolved promise and continue the boot. There wont be auth cookies, the stats image won't load, but that's all: everything else in the app will work.

File: client/desktop/lib/cookie-auth/index.js [hunk 11]
```diff
@@ -70,49 +67,62 @@ function parseCookie( cookieStr ) {
...
 			}
-			currentRequest = authorize( userData.username, token ).on(
-				'response',
-				setSessionCookies( window, onAuthorized )
-			);
 		} else {
-			// retrieve all cookies
+			log.info( 'No user data, clearing session cookies...' );
+
 			window.webContents.session.cookies.get( {}, function ( e, cookies ) {
```
Comment (jsnajdr) [inline client/desktop/lib/cookie-auth/index.js:123]: `session.cookies.get` also returns a promise in recent Electrons, so the callback can be removed in favor of an `await`.

File: client/desktop/lib/cookie-auth/index.js (hunk 6 above)
Comment (jsnajdr) [reply client/desktop/lib/cookie-auth/index.js:99]: Spawning several parallel async "threads" in `forEach`, waiting for all of them to finish and resolving then... that's exactly what `Promise.all` does. The `cookieHeaders.forEach` cycle can be rewritten as:
```js
await Promise.all( cookieHeaders.map( async ( cookieStr ) => {
  // ...
} ) );
```

And `setSessionCookies` can become an `async` function.

File: client/desktop/lib/cookie-auth/index.js [hunk 12]
```diff
@@ -70,49 +67,62 @@ function parseCookie( cookieStr ) {
...
-function setSessionCookies( window, onComplete ) {
-	return function ( response ) {
-		let cookieHeaders = response.headers[ 'set-cookie' ];
+function setSessionCookies( window, authorizeResponse ) {
+	return new Promise( ( resolve ) => {
+		let cookieHeaders = authorizeResponse.headers[ 'set-cookie' ];
 		let count = 0;
 		if ( ! Array.isArray( cookieHeaders ) ) {
 			cookieHeaders = [ cookieHeaders ];
 		}
```
Comment (jsnajdr) [inline client/desktop/lib/cookie-auth/index.js:76]: I'm wondering what this code does when there are no 'set-cookies' headers in the response. Then `cookieHeaders` are `undefined`, are converted to `[ undefined ]` and then `parseCookie` crashed when trying to call `.split()` on `undefined` ? I didn't test it, but I suspect there is a pre-existing bug.
